### PR TITLE
`ultralytics 8.3.76` fix `dynamic` batch inference with NMS export

### LINKED ISSUE #19243: Run  failed when using nms in batch infer
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Export

### Bug

I want to convert the model to ONNX format and integrate NMS into the model. However, after conversion, when I try to perform batch inference with dynamic shapes, I get a Split error. When I change the batch size during conversion, the error numbers also change accordingly. It only works correctly when the preset batch size matches the inference batch size. Is this because some tensor values were converted to constants somewhere? I hope to get an answer.

err message:
```
[E:onnxruntime:, sequential_executor.cc:516 ExecuteKernel] Non-zero status code returned while running Split node. Name:'/Split_2' Status Message: Cannot split using values in 'split' attribute. Axis=0 Input shape={3,8400} NumOutputs=1 Num entries in 'split' (must equal number of outputs) was 1 Sum of sizes in 'split' (must equal size of selected axis) was 1
```
the split node:
Image

### Environment

ultralytics 8.3.74
torch 2.4.0
onnx 1.17.0
onnxruntime-gpu 1.19.0
opencv-python  4.10.0.84
python 3.12.4

### Minimal Reproducible Example

export code:
```
from ultralytics import YOLO
import cv2
import numpy as np

model = YOLO('yolov11n-face.pt')  # load a pretrained YOLOv8n detection model
image = cv2.imread('../datasets/face_test/image.jpg')
results = model((image, image2, image3, image4))  # predict on an image

model.export(format="onnx", nms=True, dynamic=True, simplify=True)
```
test code:
```
image = cv2.imread('../datasets/face_test/image1.jpg')
image2 = cv2.imread('../datasets/face_test/image2.jpg')
results = model((image, image2))  # predict on an image

output_path = 'yolov11n-face.onnx'

image_list = [image, image2]
w = torch.Tensor([i.shape[0] for i in image_list]).to(model.device)
h = torch.Tensor([i.shape[1] for i in image_list]).to(model.device)
in_tensor = preprocess(image_list) # (2, 640, 640, 3)

onnx_model = onnx.load(output_path)
onnx.checker.check_model(onnx_model)
ort_session = onnxruntime.InferenceSession(output_path)
ort_inputs = {
    ort_session.get_inputs()[0].name: to_numpy(in_tensor.cpu())
}
# print(ort_inputs)
for _ in range(1):
    ort_outs = ort_session.run(None, ort_inputs)

print('onnx out:')
print(ort_outs)
```

### Additional

_No response_

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

### PR DESCRIPTION
**Pull request by Y-T-G**
Closes https://github.com/ultralytics/ultralytics/issues/19243

Currently, with `dynamic=True` and `nms=True`, dynamic batch sizes don't work. They are fixed to the `batch` used during export. Getting for loop to be dynamic is difficult with tracing export. It gets unrolled during tracing.

This PR uses a workaround by padding the `pred` to the batch size used during export. This means the max batch size is fixed (similar to how TensorRT works), but dynamic batch inputs still work. If the input has more than max `batch` images, the preds for the extra images would be 0.

```bash
yolo export model=yolo11n.pt format=onnx nms=True dynamic=True batch=4
yolo predict model=yolo11n.onnx
```

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary  
Batched export now dynamically adjusts padding for consistent outputs, improving runtime flexibility and preventing errors. 🚀

### 📊 Key Changes  
- Updated the logic in export handling to dynamically adjust batch padding when `dynamic` is enabled and batch size exceeds inputs.  
- Simplified code and reduced redundancy, e.g., reuse of `kwargs` for tensor initialization.  
- Fixed shape handling to ensure outputs are correct for segmentation tasks or when `max_det` is reached.  

### 🎯 Purpose & Impact  
- **🛠️ Robustness**: Avoids errors caused by inconsistent input sizes when exporting models in batched scenarios.  
- **🚀 Flexibility**: Supports dynamic batching configurations more effectively, allowing custom batch sizes.  
- **🔧 Improved Usability**: Users can perform batch inference and export with fewer concerns about size mismatches, enhancing experience and reliability.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Y-T-G, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! To ensure a smooth review and integration process, please take a look at the checklist below and ensure everything is in order:

- ✅ **Define a Purpose**: Clearly explain the purpose of your PR in the description. You’ve done a great job linking to the relevant issue [#19243](https://github.com/ultralytics/ultralytics/issues/19243). If there are any updates, be sure to add them for context.
- ✅ **Synchronize with Source**: Ensure your branch is synchronized with the `ultralytics/ultralytics` `main` branch. If needed, click the 'Update branch' button or run `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Confirm all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks pass. If any checks fail, make sure to address them before requesting a review.
- ✅ **Enhance Documentation**: If your changes impact functionality, ensure the relevant [documentation](https://docs.ultralytics.com/) is updated.
- ✅ **Include Tests**: If applicable, include or update the tests to cover these changes, and confirm all tests pass. This helps maintain a high level of code reliability and coverage.
- ✅ **Sign the CLA**: If this is your first contribution to Ultralytics, please make sure you’ve signed our [Contributor License Agreement (CLA)](https://docs.ultralytics.com/help/CLA/) by commenting, "I have read the CLA Document and I sign the CLA."
- ✅ **Provide Minimal Changes**: Ensure the changes in your PR are focused and limited to only those necessary for addressing the issue. “Hack away the unessential” — Bruce Lee 🌟

For your specific PR:
- I noticed that your solution adapts the export logic to better handle dynamic batch sizes with NMS, which is a useful enhancement. 💡 If possible, provide a **minimum reproducible example (MRE)** for testing your changes in various export and inference scenarios (e.g., specific scripts or commands that validate dynamic batching behavior).
- Thank you for including a detailed diff and explanation, which is immensely helpful during the review process. 🙌

For additional guidelines or inspiration, check out our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). This is an automated response to assist you in navigating the PR process; an Ultralytics engineer will review your code and provide further assistance shortly.

Thank you for contributing to Ultralytics and improving our tools for the community! 🚀✨

**Comment by UltralyticsAssistant** [comment]
✨ Amazing work, @Y-T-G and @glenn-jocher! This PR is a fantastic step forward in making our exports smarter and more adaptable. The dynamic padding enhancement aligns perfectly with our mission to stay flexible and reliable for our users. 🚀  

As Leonardo da Vinci once said, "Simplicity is the ultimate sophistication." Your efforts to simplify the code while boosting its robustness and usability embody this principle beautifully.  

Thank you for your dedication to improving Ultralytics—this contribution truly makes a difference! 💡👏

**Comment by glenn-jocher** [comment]
This is a great fix for a common issue with batch handling in exports. The approach of padding predictions to the max batch size is smart and aligns with how TensorRT handles dynamic batches.

Looking at the implementation in the exporter, this change maintains compatibility with various export formats while enabling true dynamic batching during inference. I particularly like the elegant solution using `torch.max(torch.tensor(self.args.batch - bs), torch.tensor(0))` to conditionally add padding only when needed.

For users who encounter this in the future, this means you can now successfully:
1. Export with `dynamic=True` and `nms=True` together
2. Set a maximum batch size during export (e.g., `batch=4`)
3. Run inference with varying batch sizes up to that maximum

Great improvement to the exporter functionality!
